### PR TITLE
txn: Pipelined DML: sort and remove duplicated keys when buffer batch get

### LINKED ISSUE #16882: some results missing when keys of buffer batch get are unsorted
## Bug Report

ref https://github.com/tikv/tikv/issues/16291

### What version of TiKV are you using?

b6a029e3c0ddd22d5259efed3861e3f3f489e425

### Steps to reproduce

Bug in TiDB side pingcap/tidb#52692

Can also run this case

```rust
#[test_case(test_raftstore::must_new_cluster_and_kv_client)]
#[test_case(test_raftstore_v2::must_new_cluster_and_kv_client)]
fn test_pipelined_dml_buffer_get_unordered_keys() {
    let (_cluster, client, ctx) = new_cluster();
    let keys = vec![
        b"key1".to_vec(),
        b"key2".to_vec(),
        b"key3".to_vec(),
        b"key4".to_vec(),
    ];

    // flushed lock can be observed by another read
    let mut req = FlushRequest::default();
    req.set_mutations(
        keys.iter()
            .map(|key| Mutation {
                op: Op::Put,
                key: key.clone(),
                value: key.clone(),
                ..Default::default()
            })
            .collect::<Vec<_>>()
            .into(),
    );
    req.set_generation(1);
    req.set_context(ctx.clone());
    req.set_start_ts(1);
    req.set_primary_key(keys[0].clone());
    let resp = client.kv_flush(&req).unwrap();
    assert!(!resp.has_region_error());
    assert!(resp.get_errors().is_empty());

    let mut reversed_keys = keys.clone();
    reversed_keys.reverse();
    let duplicated_keys = keys
        .clone()
        .iter()
        .flat_map(|key| vec![key.clone(), key.clone()])
        .collect();
    let cases = vec![keys.clone(), reversed_keys, duplicated_keys];
    for case in cases {
        let mut req = BufferBatchGetRequest::default();
        req.set_keys(case.into());
        req.set_context(ctx.clone());
        req.set_version(1);
        let resp = client.kv_buffer_batch_get(&req).unwrap();
        let pairs = resp.get_pairs();
        assert_eq!(pairs.len(), 4);
        let pairs_map = pairs
            .iter()
            .map(|pair| (pair.get_key().to_vec(), pair.get_value().to_vec()))
            .collect::<HashMap<_, _>>();
        for key in &keys {
            assert_eq!(pairs_map.get(key).unwrap(), key.as_slice());
        }
    }
}
```

### What did you expect?

Both SQL and KV tests pass.

### What did happened?

Test failed

### PR DESCRIPTION
**Pull request by you06**
### What is changed and how it works?

Issue Number: Close #16882

What's Changed:

```commit-message
The PR fixes 2 problems, unsorted keys can get wrong results and repeat keys,
and repeat keys can get duplicate results.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -999,6 +999,15 @@ impl<E: Engine, L: LockManager, F: KvFormat> Storage<E, L, F> {
             )
         });
         let priority_tag = get_priority_tag(priority);
+        keys.sort();
+        let mut i = 1;
```
Comment (cfzjywxk) [inline src/storage/mod.rs:1003]: Why not use the `dedup` interface of `Vec`? The current implementation is O(n^2) time complexity.

**Review by ekexium — APPROVED** [review]

**Comment by ekexium** [comment]
Or we can set scan_mode to `None`?

**Comment by you06** [comment]
> Or we can set scan_mode to `None`? It must be a mistake to use `Forward`.

With key sorts, `Forward` mode also guarantees the correctness, the performance is the problem.

I checked the stats of `Forward`, it takes 3 nexts and 1 seek to serve `buffer_batch_get` 4 keys, I think the performance of `Forward` may be better in this case.

With further investigation and experiments, this might also be a possible optimization for `batch_get`.

Change the  it to `None` and leave a TODO in this PR.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**